### PR TITLE
Load selected lines from query params

### LINKED ISSUE #843: Load selected lines from query params
See https://github.com/carbon-app/carbon/issues/824#issuecomment-513694282 for context

- [x] Should use existing state/routing system

### PR DESCRIPTION
**Pull request by nemesv**
Closes #843

### CODE REVIEW

File: lib/routing.js [hunk 1]
```diff
@@ -72,7 +72,8 @@ const readMappings = [
   { field: 'readonly', type: 'bool' },
   { field: 'id' },
   { field: 'highlights', type: 'parse' },
-  { field: 'code', type: 'decode' }
+  { field: 'code', type: 'decode' },
+  { field: 'selectedLines' }
```
Comment (mfix22) [inline lib/routing.js]: Can se do `sl:selectedLines`?

File: components/Carbon.js [hunk 2]
```diff
@@ -377,13 +376,21 @@ function useHighlightLoader() {
...
+  { type, lineNumber, numLines, selectedLines }
+) {
   const newState = {}
 
   if (type === 'GROUP' && prevLine) {
     for (let i = Math.min(prevLine, lineNumber); i < Math.max(prevLine, lineNumber) + 1; i++) {
       newState[i] = selected[prevLine]
     }
+  }
+  if (type === 'MULTILINE') {
```
Comment (mfix22) [inline components/Carbon.js]: Hmm what is `MULTILINE` used for? Just for selecting  any arbitrary list of lines? 🙂

File: components/Carbon.js [hunk 3]
```diff
@@ -402,27 +409,34 @@ function selectedLinesReducer({ prevLine, selected }, { type, lineNumber, numLin
...
-  }, [state.selected, props.children, props.config])
+  }, [state.selected, props.children, props.config, editorRef])
 
-  return React.useCallback(function onGutterClick(editor, lineNumber, gutter, e) {
-    editorRef.current = editor
+  React.useEffect(() => {
+    let selectedLines = (props.config.selectedLines || '')
+      .split(',')
+      .map(i => parseInt(i))
+      .filter(i => !isNaN(i))
```
Comment (mfix22) [inline components/Carbon.js]: This number filtering/mapping should happen during the query param parsing here: https://github.com/carbon-app/carbon/pull/878/files#diff-7bd084e1642aa17e0074c18f5a25a0aeR76

File: components/Carbon.js (hunk 2 above)
Comment (nemesv) [reply components/Carbon.js]: Yes, that was my idea to support selecting a list of lines. Do you have a better suggestion for the name? In theory the other two actions GROUP and LINE could be also expressed as MULTILINE but I didn't want to change to much.

File: components/Carbon.js (hunk 3 above)
Comment (nemesv) [reply components/Carbon.js]: sure, I will move it

File: lib/routing.js (hunk 1 above)
Comment (nemesv) [reply lib/routing.js]: Ok, I will rename it

File: components/Carbon.js [hunk 4]
```diff
@@ -402,27 +409,31 @@ function selectedLinesReducer({ prevLine, selected }, { type, lineNumber, numLin
   }
 }
 
-function useGutterClickHandler(props) {
-  const editorRef = React.useRef(null)
+function useSelectedLines(props, editorRef) {
   const [state, dispatch] = React.useReducer(selectedLinesReducer, {
     prevLine: null,
     selected: {}
   })
```
Comment (mfix22) [inline components/Carbon.js]: @nemesv quicks thoughts. Now that we are also story `selectedLines` in the `Editor` state, we can probably remove this local state and just call something like `this.props.setSelectedLines` instead. That way we don't have this state in two places. 

We _don't_ need it in this PR, but do you think that sounds feasible?

**Comment by mfix22** [comment]
@nemesv this is looking really good! I will test it out more tomorrow and try and get it merged.

**Comment by mfix22** [comment]
@nemesv sorry about this, but can you fix the merge conflicts? Feel free to just check out `master` and then re add yourself as a contributor.

Alternatively, I will add you as a contributor once we merge this automatically!

**Comment by mfix22** [comment]
@nemesv this is awesome! Great work! Really appreciate your help 👍

**Pull request merged**